File: android/app/src/main/kotlin/com/cruisemesh/app/chat/DraftStore.kt
package com.cruisemesh.app.chat

import android.content.Context
import android.util.Base64

object DraftStore {
    private const val PREFS = "cruisemesh_chat_drafts"

    fun load(context: Context, chatId: ByteArray): String =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).getString(key(chatId), "").orEmpty()

    fun save(context: Context, chatId: ByteArray, text: String) {
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val previous = prefs.getString(key(chatId), "").orEmpty()
        val edit = prefs.edit()
        if (text.trim('\n', '\r').isEmpty()) edit.remove(key(chatId)) else edit.putString(key(chatId), text)
        edit.apply()
        // XP1: don't fire chat-changed on every keystroke -- the chat list re-reads
        // the live draft text on its own reload passes, so only the presence
        // transition (a draft appearing/disappearing) needs to be announced.
        if (DraftChangeSignal.shouldNotify(previous, text)) {
            ChatEvents.notifyChatChanged(chatId)
        }
    }

    private fun key(chatId: ByteArray): String =
        Base64.encodeToString(chatId, Base64.NO_WRAP or Base64.URL_SAFE)
}

File: ios/CruiseMesh/Chat/DraftStore.swift
import Foundation

enum DraftStore {
    private static let prefix = "cruisemesh.chat.draft."

    static func load(chatId: Data) -> String {
        AppDefaults.current.string(forKey: key(chatId)) ?? ""
    }

    static func save(chatId: Data, text: String) {
        let normalized = text.trimmingCharacters(in: .newlines)
        let previous = load(chatId: chatId)
        if normalized.isEmpty {
            AppDefaults.current.removeObject(forKey: key(chatId))
        } else {
            AppDefaults.current.set(text, forKey: key(chatId))
        }
        // XP1: don't fire chat-changed (and, via ChatView's own sink, a
        // reload) on every keystroke -- only the presence transition (a
        // draft appearing/disappearing) needs to be announced.
        if DraftChangeSignal.shouldNotify(previous: previous, next: text) {
            ChatEvents.notifyChatChanged(chatId)
        }
    }

    private static func key(_ chatId: Data) -> String {
        prefix + chatId.base64EncodedString()
    }
}
